src/full/Agda/TypeChecking/Monad/Base.hs

```diff
@@ -2677,12 +2728,491 @@ ifTopLevelAndHighlightingLevelIs ::
+
+#if __GLASGOW_HASKELL__ >= 804
+{-# COMPLETE H'LHS #-}
+{-# COMPLETE H'RHS #-}
+{-# COMPLETE H'Compat #-}
+{-# COMPLETE H'Both #-}
+#endif
+
+type OnSide = Het
+type OnLHS = OnSide 'LHS
+type OnRHS = OnSide 'RHS
+
+pattern OnLHS :: a -> OnLHS a
+pattern OnLHS a = Het a
+pattern OnRHS :: a -> OnRHS a
+pattern OnRHS a = Het a
+
+#if __GLASGOW_HASKELL__ >= 804
+{-# COMPLETE OnLHS #-}
+{-# COMPLETE OnRHS #-}
+#endif
+
+onSide :: forall s a m b. (MonadAddContext m, HetSideIsType s) => (a -> m b) -> Het s a -> m (Het s b)
+onSide κ = switchSide @s . traverse κ
+
+onSide_ :: forall s a m b. (MonadAddContext m, HetSideIsType s) => (a -> m b) -> Het s a -> m b
+onSide_ κ = fmap (unHet @s) . switchSide @s . traverse κ
+
+-- Dependent type boilerplate
+data instance SingT (a :: HetSide) where
+  SLHS    :: SingT 'LHS
+  SRHS    :: SingT 'RHS
+  SCompat :: SingT 'Compat
+  SBoth   :: SingT 'Both
+instance Sing 'LHS    where sing = SLHS
+instance Sing 'RHS    where sing = SRHS
+instance Sing 'Both   where sing = SBoth
+instance Sing 'Compat where sing = SCompat
+
+-- | Distinguishes which sides of a twin type corresponds to a single type
+type family HetSideIsType_ (s :: HetSide) :: Bool where
+  HetSideIsType_ 'LHS    = 'True
+  HetSideIsType_ 'RHS    = 'True
+  HetSideIsType_ 'Compat = 'True
+  HetSideIsType_ 'Both   = 'True
+type HetSideIsType s = (Sing s, HetSideIsType_ s ~ 'True)
+
+-- | Distinguishes which sides of a twin type corresponds to a single type
+type family LeftOrRightSide_ (s :: HetSide) :: Bool where
+  LeftOrRightSide_ 'LHS    = 'True
+  LeftOrRightSide_ 'RHS    = 'True
+  LeftOrRightSide_ 'Compat = 'False
+  LeftOrRightSide_ 'Both   = 'False
+type LeftOrRightSide s = (Sing s, LeftOrRightSide_ s ~ 'True)
+
+-- | Distinguishes which sides of a twin type corresponds to a single type
+type family AreSides_ (s₁ :: HetSide) (s₂ :: HetSide) :: Bool where
+  AreSides_ 'LHS 'RHS = 'True
+  AreSides_ 'RHS 'LHS = 'True
+  AreSides_ _    _    = 'False
+type AreSides s₁ s₂ = (LeftOrRightSide s₁, LeftOrRightSide s₂, AreSides_ s₁ s₂ ~ 'True)
+
+newtype Het (side :: HetSide) t = Het { unHet :: t }
+  deriving (Foldable, Traversable, Pretty)
+
+deriving instance (Typeable side, Data t) => Data (Het side t)
+deriving instance Show t => Show (Het side t)
+deriving instance Functor (Het side)
+
+deriving instance AllMetas a => AllMetas (Het side a)
+deriving instance Sized a => Sized (Het side a)
+
+instance Decoration (Het s) where
+  traverseF f = fmap coerce . f . coerce
+  distributeF = traverseF id
+
+-- | Converse of `distributeF` for `Het`
+pullHet :: Coercible (f (Het side a)) (f a) => f (Het side a) -> Het side (f a)
+pullHet = coerce
+
+instance Applicative (Het s) where
+  pure = Het
+  Het f <*> Het a = Het (f a)
+
+instance Monad (Het s) where
+  Het a >>= f = f a
+
+instance (HetSideIsType side, Free a) => Free (Het side a) where
+  freeVars' (Het a) = freeVars' a
+
+instance TermLike a => TermLike (Het side a) where
```

Comment:
This could probably be derived via `GeneralizedNewtypeDeriving` (and some others as well).